### ISSUE TITLE
Improve ncap2 command-line string handling in Windows

### ISSUE THREAD

**Issue by czender** [nco/nco#91]
Basically, it has never worked. Please create a regression test whose failure demonstrates that, and try to improve matters. Goal is for ncap2 on Windows to work identically to Linux, or, failing that, at least to work better with command line strings so `ncap2 -s 'foo=bar' in.nc out.nc` works on Windows instead of requiring` ncap2 -S foo.nco in.nc out.nc`.

**Comment by hmb1** [comment]
finally managed to compile NCO on windows;  
looking at the command line:   

`ncap2 -v  -s "bbbb=one;"   ..\data\in.nc foo.nc`

...Henry

**Comment by ocefpaf** [comment]
What we have in 

https://github.com/nco/nco/blob/master/conda.recipe/run_test.bat#L7

seems to work:

https://ci.appveyor.com/project/czender/nco/build/1.0.276#L1007

**Comment by czender** [comment]
Yes, I think we removed all the Windows ncap2 tests that failed so we could pass Appveyor tests. They all involved command line strings. Apparently some commands do work, as you show.

**Comment by ocefpaf** [comment]
Ah OK. I did not recall that we removed some commands.

**Comment by hmb1** [comment]
@ocefpaf 
@pedro-vicente 

Have created the branch **fix-win-ncap2**    where I am putting the windows ncap2 fixes.
It looks like some of the issues are due to subtle differences between  the LINUX STL vector implementation and the Microsoft one .

...Henry

**Comment by hmb1** [comment]
@pedro-vicente 
some of the issues are due to   differences between  the antr2  string **c_str()** method and the std::string **c_str()** method.  for example.

 ```
std::string s_var_nm;
s_var_nm = vtr_args[0]->getText()

/ / this works   
Nvar=prs_arg->var_vtr.find(s_var_nm.c_str());

// this doesn't work
Nvar=prs_arg->var_vtr.find(vtr_args[0]->getText().c_str());
```
 not sure if this can be fixed in antlr2 or go throu the and rework the ncap2 code ?

...Henry

**Comment by hmb1** [comment]
@czender 
@pedro-vicente 

The following ncap2  now scripts run correctly
`ncap2 -v -O -S  gsl_sf.in in.nc sf.nc`   
`ncap2 -v -O -S bsc_tst.nco in.nc bsc-tst.nc`

Udunits is not working correctly 
`ncap2 -v -O -S tst-udunits.nco in.nc ud.nc`

when I remove the statement  
`bfr_free=(char *)nco_free(bfr_free); `  from  `nco_cln_prs_tm()`
 the script runs. So it seems there is some issue when passing a pointer  to the udunits2 library
 The libexpat library is  x86  and  ncap2 is x64. It also seems that the issues with c_str() are related to passing pointers from antlr back to ncap2.  I am pretty much at the limit of my knowledge 

...Henry

**Comment by czender** [comment]
In this situation, and situations like this, I suggest placing the offending code,
`bfr_free=(char *)nco_free(bfr_free);` within #ifndef MSVC so that Linux frees the memory, and Windows does not. Within that #ifdef you also place explanatory notes that describe why Windows is not allowed to free the memory, the date, and pointer to this GitHub URL so future code debuggers can start from where you left off without re-discovering all of this. Thank you.

**Comment by pedro-vicente** [comment]
oopsy , I just saw these messages from the github ticket opened 19 days ago .

I use for all github things this email
me@pedro-vicente.net

and since I am "watching" many projects I get lots of email (and some folk spend all day githubbing messages)
so, if you please send a personal email to my usual email in the future when you do not get an immediate reply...

anyway, I'll look into this 

if you are curious about the pedro-vicente.net domain, here it is in its full one page glory

http://pedro-vicente.net/

**Comment by pedro-vicente** [comment]
@hmb1 
>>> The libexpat library is x86 and ncap2 is x64

this cannot be done
the cmake/bld.bat script makes sure that all dependencies are built with x64 (or x86) with this line

`set` MSVC_VERSION="Visual Studio 14 2015 Win64"

It seems you were able to build on Windows? can you give it a try?

**Comment by hmb1** [comment]
hi Pedro, attached are my  expat.dll  and udunits2 dll ...Henry

On 21/06/2018 at 3:46 AM, "Pedro Vicente"  wrote:

	@hmb1    

	The libexpat library is x86 and ncap2 is x64    

	this cannot be done
 the cmake/bld.bat script makes sure that all dependencies are built
with x64 (or x86) with this line 

	set MSVC_VERSION="Visual Studio 14 2015 Win64" 

	It seems you were able to build on Windows? can you give it a try?

**Comment by czender** [comment]
What is the status of this? Is there any prospect of fixing the remaining issues?

**Comment by hmb1** [comment]
Hi Charlie,  Can you give Pedro a prod on this one  !?   The branch 
win-fix-ncap2  is getting stale 
....Henry

On 23/08/2018 at 11:53 PM, "Charlie Zender"  wrote:

	What is the status of this? Is there any prospect of fixing the
remaining issues?

**Comment by pedro-vicente** [comment]
in windows i get an exception
error stack is


```
 	ncap2.exe!std::vector<int,std::allocator<int> >::operator[](unsigned __int64 _Pos) Line 1219	C++
>	ncap2.exe!antlr::CircularQueue<int>::elementAt(unsigned __int64 idx) Line 45	C++
 	ncap2.exe!antlr::InputBuffer::LA(unsigned int i) Line 82	C++
 	ncap2.exe!antlr::CharScanner::LA(unsigned int i) Line 548	C++
 	ncap2.exe!ncoLexer::mBLASTOUT(bool _createToken) Line 483	C++
 	ncap2.exe!`ncoLexer::nextToken'::`1'::catch$7() Line 393	C++
 	[External Code]	
 	ncap2.exe!ncoLexer::nextToken() Line 289	C++
 	ncap2.exe!antlr::TokenStreamSelector::nextToken() Line 56	C++
 	ncap2.exe!antlr::TokenBuffer::fill(unsigned int amount) Line 50	C++
 	ncap2.exe!antlr::TokenBuffer::LA(unsigned int i) Line 58	C++
 	ncap2.exe!antlr::LLkParser::LA(unsigned int i) Line 47	C++
 	ncap2.exe!ncoParser::program() Line 40	C++
 	ncap2.exe!parse_antlr(std::vector<prs_cls,std::allocator<prs_cls> > & prs_vtr, char * fl_spt_usr, char * cmd_ln_sng) Line 183	C++
 	ncap2.exe!main(int argc, char * * argv) Line 854	C++
 	[External Code]	
```

this call

```
inline T elementAt( size_t idx ) const
	{
		return storage[idx+m_offset];
	}
```

has and invalid idx value , size_t(-1)

**Comment by hmb1** [comment]
IS this with the new branch "win-fix-ncap2" ?

**Comment by pedro-vicente** [comment]
First I tried master, then "win-fix-ncap2", same results...
what are the additions of "win-fix-ncap2"?

**Comment by hmb1** [comment]
Hi Pedro,   
your tests 

```
ncks -v area ../data/in_grp.nc out.nc
ncap2 -v -s 'area[lat]=100' out.nc out1.nc
```
 worked for me on the branch **win-fix-ncap2**   with a build on Visual Studio community 2017.
Note that ncap2 cannot handle groups - It will only read  from the root group

**Comment by pedro-vicente** [comment]
>>> Note that ncap2 cannot handle groups - It will only read from the root group

right, bad example. I tried with in.nc

```
ncks -v area ../data/in.nc out.nc
ncap2 -v -s 'area[lat]=100' out.nc out1.nc
```
and I get the same error in both branches.
but using double quotes has no problem

```
ncap2 -v -s "area[lat]=100" out.nc out1.nc
```
 I guess a simple fix would be to replace single quote ' for double quote "  in windows

**Comment by pedro-vicente** [comment]
> I guess a simple fix would be to replace single quote ' for double quote " in windows
this done programmatically, not just from the command line

**Comment by hmb1** [comment]
@pedro-vicente 
if you wanna concentrate on the  dll /crt  - I can  do the single quote thing in ncap2

**Comment by pedro-vicente** [comment]
@hmb1 currently it has static CRT already. This README.mt shows the steps, some files have to be manually edited because the maintainers of some dependencies like HDF5 don't want to make the changes to static CRT in their scripts

https://github.com/nco/nco/blob/master/cmake/README.md

**Comment by czender** [comment]
What is the status of this? It seems like it was near resolution by replacing single with double quotes in Windows. Or perhaps that's not necessary? Someone please clarify the remaining Windows incompatibilities.
